builtInFunctions/upgradeNFTAttributes.go

```diff
@@ -0,0 +1,135 @@
+package builtInFunctions
+
+import (
+	"math/big"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-vm-common"
+	"github.com/ElrondNetwork/elrond-vm-common/atomic"
+	"github.com/ElrondNetwork/elrond-vm-common/check"
+)
+
+type esdtNFTupdate struct {
+	*baseEnabled
+	keyPrefix    []byte
+	marshalizer  vmcommon.Marshalizer
+	pauseHandler vmcommon.ESDTPauseHandler
+	rolesHandler vmcommon.ESDTRoleHandler
+	gasConfig    vmcommon.BaseOperationCost
+	funcGasCost  uint64
+	mutExecution sync.RWMutex
+}
+
+// NewESDTNFTAddUriFunc returns the esdt NFT update attribute built-in function component
+func NewESDTNFTUpdateAttributesFunc(
+	funcGasCost uint64,
+	gasConfig vmcommon.BaseOperationCost,
+	marshalizer vmcommon.Marshalizer,
+	pauseHandler vmcommon.ESDTPauseHandler,
+	rolesHandler vmcommon.ESDTRoleHandler,
+	activationEpoch uint32,
+	epochNotifier vmcommon.EpochNotifier,
+) (*esdtNFTupdate, error) {
+	if check.IfNil(marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(pauseHandler) {
+		return nil, ErrNilPauseHandler
+	}
+	if check.IfNil(rolesHandler) {
+		return nil, ErrNilRolesHandler
+	}
+
+	e := &esdtNFTupdate{
+		keyPrefix:    []byte(vmcommon.ElrondProtectedKeyPrefix + vmcommon.ESDTKeyIdentifier),
+		marshalizer:  marshalizer,
+		funcGasCost:  funcGasCost,
+		mutExecution: sync.RWMutex{},
+		pauseHandler: pauseHandler,
+		gasConfig:    gasConfig,
+		rolesHandler: rolesHandler,
+	}
+
+	e.baseEnabled = &baseEnabled{
+		function:        vmcommon.BuiltInFunctionESDTNFTAddURI,
+		activationEpoch: activationEpoch,
+		flagActivated:   atomic.Flag{},
+	}
+
+	epochNotifier.RegisterNotifyHandler(e)
+
+	return e, nil
+}
+
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtNFTupdate) SetNewGasConfig(gasCost *vmcommon.GasCost) {
+	if gasCost == nil {
+		return
+	}
+
+	e.mutExecution.Lock()
+	e.funcGasCost = gasCost.BuiltInCost.ESDTNFTAddQuantity
+	e.gasConfig = gasCost.BaseOperationCost
+	e.mutExecution.Unlock()
+}
+
+// ProcessBuiltinFunction resolves ESDT NFT add quantity function call
+// Requires 3 arguments:
+// arg0 - token identifier
+// arg1 - nonce
+// arg2 - new attributes
+func (e *esdtNFTupdate) ProcessBuiltinFunction(
+	acntSnd, _ vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+	e.mutExecution.RLock()
+	defer e.mutExecution.RUnlock()
+
+	err := checkESDTNFTCreateBurnAddInput(acntSnd, vmInput, e.funcGasCost)
+	if err != nil {
+		return nil, err
+	}
+	if len(vmInput.Arguments) != 3 {
+		return nil, ErrInvalidArguments
+	}
+
+	err = e.rolesHandler.CheckAllowedToExecute(acntSnd, vmInput.Arguments[0], []byte(vmcommon.ESDTRoleNFTUpdateAttributes))
+	if err != nil {
+		return nil, err
+	}
+
+	gasCostForStore := uint64(len(vmInput.Arguments[2])) * e.gasConfig.StorePerByte
```

Comment:
this gas cost check is somehow duplicated code as found in other builtin functions. Consider having this extracted in a common function.

Reply:
every built in function has his own logic of gas handling and his own gas costs.